### PR TITLE
GH-2727 simplified path expression processing and fixed nested inverse optional handling

### PR DESCRIPTION
**Pull request by jeenbroekstra**
GitHub issue resolved: #2727  

Briefly describe the changes proposed in this PR:

- refactor of property path processing to a cleaner tree-visitor approach instead of a massive if-else. 
- simplified path modifier processing code (old code was still written for {min,max} syntax which is no longer supported)
- added regression test for GH-2727

To explain the main ideas behind the refactor:

A SPARQL path expression is parsed into an AST tree structure. Each path expression is a PathAlternative, consisting of one more PathSequences. Each PathSequence, in turn, consists of one or more PathElements. PathElements have optional modifiers (?, + and *), and can be either negated or inverted (or both). Note that a PathElement itself can be a nested path expression again.

In the previous code, the entire path expression processing was handled in a single visitor method, the `visit(PathSequence)`.  It kept internal state, manually inspected its child elements (instead of letting the visitor pattern handle it), and was generally a mess of if-else conditions and special-case handling. What it also did was manipulate the global GraphPattern field directly, instead of just returning the produced TupleExpr. 

The refactor splits this out, so that processing logic is more local to the specific element: path sequences deal with introducing intermediate variables, keeping the context of what the "current" start and end variables of the sequence are. The details of processing each element of the path are now in `visit(PathElt)`. This checks if the element is nested (and if so defers back to the visitor to handle the nested expression), or negated, or inverse. It also handles all path modifiers. 

The handling of path modifiers has also been significantly changed. Previously it had convoluted logic that dealt with building up a tree of Unions to handle things like "at least 2, at most 5". This was introduced when SPARQL was still in draft and had this `{min, max}` syntax for path expressions. I removed this because the logic was hard to follow, essentially unused, and produced buggy behavior in several cases. 

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits down to one or a few meaningful commits
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [x] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

### CODE REVIEW

File: core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/TupleExprBuilder.java [hunk 1]
```diff
@@ -100,127 +99,7 @@
...
-import org.eclipse.rdf4j.query.parser.sparql.ast.ASTTz;
-import org.eclipse.rdf4j.query.parser.sparql.ast.ASTUUID;
-import org.eclipse.rdf4j.query.parser.sparql.ast.ASTUnionGraphPattern;
-import org.eclipse.rdf4j.query.parser.sparql.ast.ASTUpperCase;
-import org.eclipse.rdf4j.query.parser.sparql.ast.ASTVar;
-import org.eclipse.rdf4j.query.parser.sparql.ast.ASTYear;
-import org.eclipse.rdf4j.query.parser.sparql.ast.Node;
-import org.eclipse.rdf4j.query.parser.sparql.ast.SimpleNode;
-import org.eclipse.rdf4j.query.parser.sparql.ast.VisitorException;
+import org.eclipse.rdf4j.query.parser.sparql.ast.*;
```
Comment (jeenbroekstra) [inline core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/TupleExprBuilder.java:102]: Heh cute, we went over 99 imports so Eclipse introduced a wildcard. I'll fix.

**Comment by jeenbroekstra** [comment]
Due to the large amount of refactoring (I got rid of a ton of legacy code) the diff on this PR may be a bit hard to follow, apologies for that. It's not 100% cleaned up, but (IMO) a lot better than what it was.

**Comment by hmottestad** [comment]
It's a bit complicated for me to be honest. Had a hard time grasping the original code, so it's tricky to understand what your changes entail. That's fine though. I don't need to understand everything :P

Only question I have is the changes to the two query plans. So you know why that has happened and also if those changes affect performance?

**Comment by jeenbroekstra** [comment]
> It's a bit complicated for me to be honest. Had a hard time grasping the original code, so it's tricky to understand what your changes entail. That's fine though. I don't need to understand everything :P

Fair enough. I have added a bit more explanation to the description (see above), but I appreciate that even with that it's a fairly opaque piece of code.

> Only question I have is the changes to the two query plans. So you know why that has happened and also if those changes affect performance?

Presumably due to the way the algebra is now built up (how we're building the TupleExpr based on the GraphPattern object kept in the visitor), expressions are added to the algebra tree in a slightly different order sometimes. The testTupleQuery one simply produces a different join order in the unoptimized form, and for what it's worth the join order in the new version looks more correct to me than what was originally there (though presumably the query planner will optimize both forms to the same thing). 

As for the subquery one, that's a comparison of the plan after execution, and from what I can tell the main difference is that the filter condition has been pushed further down the tree, leading to a smaller intermediate result set (which should make it more efficient). I'm not 100% why it can now do this optimization when it couldn't before, but given that every other test we have in place greenlines I have assumed (after looking at the new plans and thinking they look conceptually correct for the query) that those query plan tests were brittle, and have just adjusted the expectation.  

I have not yet checked how any of this affects performance. I will try to at least run the parser benchmarks and the memory store query benchmarks later for a quick comparison.

**Comment by hmottestad** [comment]
Pushing the filter down further is usually better.

The test were initially designed to check that the explain feature worked as expected, and then we added some extra cases for some poorly performing queries that you later fixed. They are kinda brittle, but then again they will only break if you actually change anything that affects the query plan - which means that if these failed but no other tests failed we might need better test coverage for the query optimisers?

Do we have any specific tests that test that the filter push optimisation works as expected?

**Comment by hmottestad** [comment]
Actually. I think I remember something. Our optimisers ignore sub query plans, so since your changes affect queries before they are optimised then it would make sense that it would affect the subquery query too!

**Comment by jeenbroekstra** [comment]
I ran the SparqlParseBenchmark against this vs master branch:

master:

```
Benchmark                                        Mode  Cnt   Score   Error  Units
SPARQLParseBenchmark.complexPathExpressionQuery  avgt   40  29.138 ± 0.898  ms/op
```

This branch:

```
Benchmark                                        Mode  Cnt   Score   Error  Units
SPARQLParseBenchmark.complexPathExpressionQuery  avgt   40  32.549 ± 1.307  ms/op
```

So we're seeing a marginal performance penalty, but to be honest given that we're talking about ~4ms difference on a pathologically complex path expression, I'm not too worried about this. 

More interesting will be to run some evaluation benchmarks with queries involving path expressions. We have nothing ready-made for this unfortunately, so I'll try and come up with a non-trivial example on the datagovbe dataset.

**Comment by hmottestad** [comment]
This might help you understand that dataset if you haven't worked with DCAT before:

http://dcat.be/images/dcat-ap.png

**Comment by damyan-ognyanov** [comment]
Looks good ...

**Comment by jeenbroekstra** [comment]
I've added two benchmark queries on the memory store, using the dcat dataset from datagovbe. Performance comparison:

master:

```
Benchmark                            Mode  Cnt    Score    Error  Units
QueryBenchmark.pathExpressionQuery1  avgt   10  110.671 ± 12.691  ms/op
QueryBenchmark.pathExpressionQuery2  avgt   10    9.255 ±  1.041  ms/op
```

This branch:

```
Benchmark                            Mode  Cnt   Score    Error  Units
QueryBenchmark.pathExpressionQuery1  avgt   10  99.722 ± 16.703  ms/op
QueryBenchmark.pathExpressionQuery2  avgt   10   8.768 ±  0.207  ms/op
```

So certainly no peformance deterioration in evaluation performance, and if anything even a mild improvement (though again I think it kind of falls within the margin of error). 

Happy with this, will merge as soon as build succeeds.
